File: app/src/main/java/com/githang/gradledoc/common/Presenter.java
package com.githang.gradledoc.common;

import android.content.Context;

import java.io.IOException;

import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;

/**
 * @author 黄浩杭 (huanghaohang@parkingwang.com)
 * @version 2016-09-11
 * @since 2016-09-11
 */
public interface Presenter<O, V extends View<O>> {
    V getView();

    void request(Context context, String url);

    void request(Context context, String url, boolean forceRefresh);

    abstract class Base<O, V extends View<O>> implements Presenter<O, V> {
        private final V mView;
        private final Model<O> mModel;

        public Base(V view, Model<O> model) {
            mView = view;
            mModel = model;
        }

        @Override
        public V getView() {
            return mView;
        }

        @Override
        public void request(final Context context, final String url) {
            this.request(context, url, false);
        }

        @Override
        public void request(final Context context, final String url, final boolean forceRefresh) {
            getView().showProgressDialog();
            Observable.create(new Observable.OnSubscribe<O>() {
                @Override
                public void call(Subscriber<? super O> subscriber) {
                    try {
                        String content = null;
                        if (!forceRefresh) {
                            content = mModel.requestFromCache(url);
                        }
                        if (content == null) {
                            content = mModel.requestFromNetwork(url);
                        }
                        subscriber.onNext(mModel.handleContent(content));
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Action1<O>() {
                        @Override
                        public void call(O o) {
                            getView().onHandle(o);
                            getView().dismissProgressDialog();
                        }
                    }, new Action1<Throwable>() {
                        @Override
                        public void call(Throwable throwable) {
                            throwable.printStackTrace();
                            getView().showToast(throwable.getMessage());
                        }
                    });
        }
    }
}

File: app/src/main/java/com/githang/gradledoc/common/Presenter.kt
package com.githang.gradledoc.common

import android.content.Context
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.io.IOException

/**
 * @author 黄浩杭 (huanghaohang@parkingwang.com)
 * @version 2016-09-11
 * @since 2016-09-11
 */
interface Presenter<O, V : View<O>> {
    val view: V

    fun request(context: Context, url: String)

    fun request(context: Context, url: String, forceRefresh: Boolean)

    abstract class Base<O, V : View<O>>(override val view: V, private val model: Model<O>) : Presenter<O, V> {

        override fun request(context: Context, url: String) {
            this.request(context, url, false)
        }

        override fun request(context: Context, url: String, forceRefresh: Boolean) {
            view.showProgressDialog()
            Observable.create(Observable.OnSubscribe<O> { subscriber ->
                try {
                    var content: String? = null
                    if (!forceRefresh) {
                        content = model.requestFromCache(url)
                    }
                    if (content == null) {
                        content = model.requestFromNetwork(url)
                    }
                    subscriber.onNext(model.handleContent(content))
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ o ->
                        view.onHandle(o)
                        view.dismissProgressDialog()
                    }, { throwable ->
                        throwable.printStackTrace()
                        view.showToast(throwable.message ?: return@subscribe)
                        view.dismissProgressDialog()
                    })
        }
    }
}
